pubsub/google/cloud/pubsub_v1/subscriber/message.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2017, Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import math
+import time
+
+
+class Message(object):
+    """A representation of a single Pub/Sub message.
+
+    The common way to interact with
+    :class:`~.pubsub_v1.subscriber.message.Message` objects is to receive
+    them in callbacks on subscriptions; most users should never have a need
+    to instantiate them by hand. (The exception to this is if you are
+    implementing a custom subclass to
+    :class:`~.pubsub_v1.subscriber.consumer.BaseConsumer`.)
+
+    .. note::
+        Messages in Google Cloud Pub/Sub are opaque blobs of bytes. This
+        means that the ``data`` attribute will consistently be a
+        :class:`bytes` object. If you want a text string, you should
+        use :meth:`bytes.decode`.
+
+    Properties:
+        message_id (str): The message ID. In general, you should not need
+            to use this directly.
+        data (bytes): The data in the message.
+        attributes (dict): The attributes sent along with the message.
+        publish_time (datetime): The time that this message was originally
+            published.
+    """
+    def __init__(self, policy, ack_id, message):
```

Comment:
You can *not* pass the policy across the concurrency boundary and you can not directly invoke policy method across the concurrency boundary. The pattern I wrote in the proof-of-concept of sending callback messages allows you to invoke actions back on the policy without passing it to the message. The only thing you can share across the concurrency boundary is *plain data* and a *user-specified queue*.

Relevant part of proof-of-concept: https://gist.github.com/jonparrott/55723f4c8c7e5b0027f645c89a42110e#file-subscriber-py-L21